### ISSUE TITLE
Suggest to use keycloak and keycloak-gatekeeper as Auth layer over polynote

### ISSUE THREAD

**Issue by Baoqi** [polynote/polynote#869]
Currently, polynote doesn't have authentication/authorization layer, which is not secure. And keycloak is a popular Open Source Identity and Access Management System. So, maybe we can propose user to use keycloak as the auth layer, if needed.

  - keycloak: https://github.com/keycloak/keycloak
  - keycloak-gatekeeper: https://github.com/keycloak/keycloak-gatekeeper

Here is a detailed instructions to do so, (but I'm still not very familiar with keycloak yet, and security is always a very complicated concept. so, please just keep this as a simple test documentation)

step 1.  run keycloak server
     docker run --name keycloak -p 8180:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin quay.io/keycloak/keycloak:9.0.2

Note: the 8180 port change,  then, we can login    "http://127.0.0.1:8180" as "admin"/"admin" user

step2. create necessary user/client in keycloak

1. create a new realm, like:  polynote
2. create role, polynote_user
   - "Roles" -> "Add Row" button
3. create new users: for example: user1
   - "Users" -> "Add user" button
   - username as "user1"
   - Then, create user
   - In the newly appeared "Role Mappings", select the "polynote_user" role, add into "assigned roles"
   - In the "Credentials", set Password like "123", and Change "Temporary" to "OFF", and "Set Password"
   - you can create other users, if the newly added user need to access polynote, then, need to also assign the "polynote_user" role to that newly created user
4. create client: polynote_client
   - "Clients" -> "Create" button
   - set "polynote_client" as "Client ID"
   - "Client Protocol" keep as the default "openid-connect"
   - "Save" button
   - in the newly appeared "Settings" Tab
   - set "Access Type" to "confidential"
   - "Valid Redirect URIs" add http://192.168.199.253:3000/oauth/callback  and  don't forget to click '+' button to really add it.
   - Click "Save" button once, before change other Tab
   - In "Credentials" Tab, keep the "Client Authenticator" as "Client Id and Secret",   and remember the "Secret", like: "d8ef710b-587b-449c-a426-751ce8819513"
5. create client scope: polynote_scope
   - "Client Scopes" -> "Create" button
   - set "polynote_scope" as "Name"
   - "Save" button
   - In the newly appeared "Mappers" Tab, "Create" button
   - set "polynote_client_audience" as "Name"
   - set "Audience" as "Mapper Type"
   - search and set "polynote_client" as "Included Client Audience"
   - keep "Add to access token" as the default "ON"
   - "Save" button
6. associate client scope to client
   - "Clients" -> select "polynote_client" Edit
   - In the newly appeared "Client Scopes", select "polynote_scope", and click "Add selected"

Step3, to use keycloak-gatekeeper as a secured reverse proxy, like. (please change the IP from 192.168.199.253 to your IP.  since I run docker on Mac, I can not use 127.0.0.1 directly)

```
docker run --name keycloak-gatekeeper -p 3000:3000 -it quay.io/keycloak/keycloak-gatekeeper:9.0.2  \
    --discovery-url=http://192.168.199.253:8180/auth/realms/polynote \
    --client-id=polynote_client \
    --client-secret=d8ef710b-587b-449c-a426-751ce8819513 \
    --listen=:3000 \
    --secure-cookie=false \
    --redirection-url=http://192.168.199.253:3000 \
    --upstream-url=http://192.168.199.253:8192 \
    --enable-default-deny=true \
    --resources="uri=/*|roles=polynote_user" 
```

Then, visit http://192.168.199.253:3000/, it should redirect to keycloak login page, once login, we can use polynote securely.

**Comment by jeremyrsmith** [comment]
I think the keycloak-gateway part is more of a deployment concern, so the most we could do there is set up a docker build that uses it.

Polynote does have a pluggable identity mechanism, so conceivably a plugin could be created that would enable integration with keycloak for identity. (You might be able to use the built-in header identity provider, depending on whether keycloak-gateway can be configured to send such a header)

I think both of these things should probably be a separate project, though. The core project should be reasonably agnostic to security solutions.

**Comment by Baoqi** [comment]
> I think the keycloak-gateway part is more of a deployment concern, so the most we could do there is set up a docker build that uses it.
> 
> Polynote does have a pluggable identity mechanism, so conceivably a plugin could be created that would enable integration with keycloak for identity. (You might be able to use the built-in header identity provider, depending on whether keycloak-gateway can be configured to send such a header)
> 
> I think both of these things should probably be a separate project, though. The core project should be reasonably agnostic to security solutions.

Yes, that's right. The keycloak part should not be included in polynote, but as a optional deployment option.  Maybe we can put some tutorials (like how to integrate external identity system with polynote)

Just learned the "pluggable identity mechanism" from commit: https://github.com/polynote/polynote/commit/abcfadd8cf2cd37b34ee9d5c7b5362d0e30ccd4d

To integrate the above keycloak example, we can add the following thing to  config.yml as: 
```
security:
  auth:
    provider: header
    config:
      header: X-Auth-Userid
      permissions:
        user1: all
      allow_anonymous: false
```

Though, in current implementation, it is not very flexible, the rules are only works for authorization based on user name (or user id).  Maybe we can enhance it to also support authorization based on passed-in roles.  for example,  After signed in from keycloak, it has HTTP Header: 

X-Auth-Roles: polynote_user,polynote_admin

etc. we can change polynote's config.yml. to check whether. "X-Auth-Roles" contains "polynote_admin" role, etc.

**Comment by jeremyrsmith** [comment]
@Baoqi an `IdentityProvider` for tighter integration with keycloak would be a welcome PR! Please comment if you want to do this; we'll make an issue and assign it to you (I know we're lacking in developer docs, but we'll help on gitter)

**Comment by jeremyrsmith** [comment]
Basically you'd need to implement `IdentityProvider` - it seems like largely it would be similar to `HeaderIdentityProvider` except permissions would come from roles rather than user-based config in the YAML. And you'd need to pull roles from the headers given by keycloak. I think this is something we'd be willing to add directly to the core distribution (since we don't currently have a good general answer to authentication in the core distribution). So you wouldn't have to deal with plugin setup, just add the new identity provider to the relevant file in `META-INF` (in `polynote-server/src/resources`)

**Comment by Baoqi** [comment]
> @Baoqi an `IdentityProvider` for tighter integration with keycloak would be a welcome PR! Please comment if you want to do this; we'll make an issue and assign it to you (I know we're lacking in developer docs, but we'll help on gitter)

Hi, @jeremyrsmith , Thanks, sure, I'd like to try that issue.
